### PR TITLE
EE-737 validate_rule_data

### PR DESCRIPTION
**Pull request by gryevns**

### CODE REVIEW

File: business_rules/utils.py [hunk 1]
```diff
@@ -85,3 +86,64 @@ def params_dict_to_list(params):
...
+        for item in rule_schema.get('variables'):
+            if item.get('name') == condition.get('name'):
+                condition_field_type = item.get('field_type')
+                variable_operators = rule_schema.get('variable_type_operators', {}).get(condition_field_type, [])
+                for operators in variable_operators:
+                    if operators['name'] == condition['operator']:
+                        return True
+                raise AssertionError('Unknown operator "{}"'.format(condition['operator']))
+
+    def validate_conditions(input_conditions, rule_schema):
```
Comment (aaronolszewski) [inline business_rules/utils.py]: Can we break this function down into smaller functions? A lot going on here...

File: business_rules/utils.py [hunk 2]
```diff
@@ -85,3 +86,64 @@ def params_dict_to_list(params):
...
+        if isinstance(input_conditions, list):
+            for condition in input_conditions:
+                validate_conditions(condition, rule_schema)
+        if isinstance(input_conditions, dict):
+            keys = input_conditions.keys()
+            if 'any' in keys or 'all' in keys:
+                if len(keys) > 1:
```
Comment (aaronolszewski) [inline business_rules/utils.py:177]: What is so special about `1`?

**Review by aaronolszewski — CHANGES_REQUESTED** [review]
I find the flow of the function `validate_conditions` quite hard to follow. Breaking it down into smaller functions would be nicer

File: business_rules/utils.py (hunk 2 above)
Comment (gryevns) [reply business_rules/utils.py:177]: If there is `any` or `all` then it only makes sense in business rules to have one key

File: business_rules/utils.py [hunk 3]
```diff
@@ -85,3 +86,86 @@ def params_dict_to_list(params):
...
+            raise AssertionError('Missing "{}" key'.format('actions'))
+        if 'conditions' not in root_keys:
+            raise AssertionError('Missing "{}" key'.format('conditions'))
+
+    def validate_condition_operator(condition, rule_schema):
+        """
+        Check provided condition contains a valid operator
+        """
+        for item in rule_schema.get('variables'):
+            if item.get('name') == condition.get('name'):
```
Comment (ipeluffo) [inline business_rules/utils.py:141]: What if there isn't any variable with the same name as the condition?
It seems that the function will return `None`

File: business_rules/utils.py [hunk 4]
```diff
@@ -85,3 +86,86 @@ def params_dict_to_list(params):
...
+            raise AssertionError('Missing condition "name" key in {}'.format(condition))
+        if not hasattr(variables, condition_name):
+            raise AssertionError('Unknown condition "{}"'.format(condition_name))
+
+    def validate_condition(condition, variables, rule_schema):
+        validate_condition_name(condition, variables)
+        validate_condition_operator(condition, rule_schema)
+        method = getattr(variables, condition.get('name'))
+        params = condition.get('params', {})
+        engine._check_params_valid_for_method(method, params, method_type.METHOD_TYPE_VARIABLE)
```
Comment (ipeluffo) [inline business_rules/utils.py]: I think it makes more sense to move this function into this module and reuse it from `engine` module.

File: tests/test_utils.py [hunk 5]
```diff
@@ -219,3 +222,189 @@ def test_export_rule_data():
             },
         ],
     }
+
+
+def test_validate_rule_data_success():
+    valid_rule = {
+        'conditions': {
+            'all': [
```
Comment (ipeluffo) [inline tests/test_utils.py:230]: I'd add an embedded condition, for example:
```python
{
  "conditions": {
    "any": [
      {
        "name": "current_inventory",
        "operator": "less_than",
        "value": 5
      },
      {
        "all": [
          {
            "name": "current_month",
            "operator": "equal_to",
            "value": "December"
          },
          {
            "name": "current_inventory",
            "operator": "less_than",
            "value": 20
          }
        ]
      }
    ]
  },
  "actions": {}
}
```

File: business_rules/utils.py (hunk 3 above)
Comment (gryevns) [reply business_rules/utils.py:141]: Already checked in `validate_condition_name`?

File: business_rules/utils.py (hunk 4 above)
Comment (gryevns) [reply business_rules/utils.py]: 👍

File: tests/test_utils.py (hunk 5 above)
Comment (gryevns) [reply tests/test_utils.py:230]: 👍

File: business_rules/utils.py (hunk 3 above)
Comment (ipeluffo) [reply business_rules/utils.py:141]: it seems correct, but there is a dependency on the execution order of the two methods... 
Can you add a return value despite for the current implementation that should never happen?

Comment (gryevns) [reply business_rules/utils.py:141]: done!

**Review by ipeluffo — APPROVED** [review]
👍

File: business_rules/utils.py [hunk 6]
```diff
@@ -126,17 +153,12 @@ def validate_conditions(input_conditions, rule_schema):
                     for k, v in input_conditions.iteritems():
```
Comment (aaronolszewski) [inline business_rules/utils.py]: I see that `k` is a throw away variable here. Is it not best practise (like I have seen in Platform) to use a `_` to explicitly show this?

File: business_rules/utils.py [hunk 7]
```diff
@@ -85,3 +86,86 @@ def params_dict_to_list(params):
...
+        if isinstance(input_conditions, list):
+            for condition in input_conditions:
+                validate_conditions(condition, rule_schema)
+        if isinstance(input_conditions, dict):
+            keys = input_conditions.keys()
+            if 'any' in keys or 'all' in keys:
+                if len(keys) > 1:
+                    raise AssertionError('Expected ONE of "any" or "all" but found {}'.format(keys))
+                else:
+                    for k, v in input_conditions.iteritems():
```
Comment (aaronolszewski) [inline business_rules/utils.py]: I see that `k` is a throw away variable here. Is it not best practise (like I have seen everywhere n Platform) to use a `_` to explicitly show this?

**Review by aaronolszewski — APPROVED** [review]

File: business_rules/utils.py [hunk 8]
```diff
@@ -85,3 +86,113 @@ def params_dict_to_list(params):
...
+        Recursively check all levels of input conditions
+        """
+        if isinstance(input_conditions, list):
+            for condition in input_conditions:
+                validate_conditions(condition, rule_schema)
+        if isinstance(input_conditions, dict):
+            keys = input_conditions.keys()
+            if 'any' in keys or 'all' in keys:
+                if len(keys) > 1:
+                    raise AssertionError('Expected ONE of "any" or "all" but found {}'.format(keys))
```
Comment (aaronolszewski) [inline business_rules/utils.py:178]: Do you not want the message to specify the quantity of keys? Here you will just return a list of the keys rather than the quantity `len(keys)` given that the message explicitly states a quantity?

Comment (gryevns) [reply business_rules/utils.py:178]: yes **ONE**

Comment (aaronolszewski) [reply business_rules/utils.py:178]: I mean in the formatting
